src/task-definitions.js

```diff
@@ -0,0 +1,347 @@
+  keybindings: ['mod+enter'],
+
+  keybindingCallback() {
+    dispatcher.changeMode('command')
+    dispatcher.evaluateCell()
+  },
+})
+
+TASKS.EvaluateAndSelectBelow = new NotebookTask({
+  title: 'Evaluate and Select Below',
+  keybindings: ['shift+enter'],
+
+  keybindingCallback() {
+    if (viewModeIsEditor()) {
+      dispatcher.changeMode('command')
+      dispatcher.evaluateCell()
+      const cellBelowId = getCellBelowSelectedId()
+      if (cellBelowId !== null) {
+        dispatcher.selectCell(cellBelowId, true)
+      } else {
+      // if cellBelowId *is* null, need to add a new cell.
+        dispatcher.addCell('javascript')
+        dispatcher.selectCell(getCellBelowSelectedId(), true)
+      }
+    }
+  },
+})
+
+TASKS.MoveCellUp = new NotebookTask({
+  title: 'Move Cell Up',
+  displayKeybinding: '\u21E7 \u2191',
+  keybindings: ['shift+up'],
+
+  callback() {
+    this.props.actions.cellUp()
+  },
+  keybindingCallback(e) {
+    if (isCommandMode()) {
+      if (e.preventDefault) {
+        e.preventDefault()
+      }
+      dispatcher.cellUp()
+    }
+  },
+})
+
+
+TASKS.MoveCellDown = new NotebookTask({
+  title: 'Move Cell Down',
+  displayKeybinding: '\u21E7 \u2193',
+  keybindings: ['shift+down'],
+
+  callback() {
+    this.props.actions.changeCellType('javascript')
+  },
+  keybindingCallback(e) {
+    if (isCommandMode()) {
+      if (e.preventDefault) {
+        e.preventDefault()
+      }
+      dispatcher.cellDown()
+    }
+  },
+})
+
+
+TASKS.SelectUp = new NotebookTask({
+  title: 'Select Cell Above',
+  displayKeybinding: '\u2191',
+  keybindings: ['up'],
+
+  keybindingCallback(e) {
+    if (isCommandMode()) {
+      // e.preventDefault blocks kbd scrolling of entire window
+      if (e.preventDefault) {
+        e.preventDefault()
+      } else { // internet explorer
+        e.returnValue = false
+      }
+      const cellAboveId = getCellAboveSelectedId()
+      if (cellAboveId !== null) { dispatcher.selectCell(cellAboveId, true) }
+    }
+  },
+})
+
+TASKS.SelectDown = new NotebookTask({
+  title: 'Select Cell Down',
+  displayKeybinding: '\u2193',
+  keybindings: ['down'],
+
+  keybindingCallback(e) {
+    if (isCommandMode()) {
+      // e.preventDefault blocks kbd scrolling of entire window
+      if (e.preventDefault) {
+        e.preventDefault()
+      } else { // internet explorer
+        e.returnValue = false
+      }
+      const cellBelowId = getCellBelowSelectedId()
+      if (cellBelowId !== null) { dispatcher.selectCell(cellBelowId, true) }
+    }
+  },
+})
+
+
+TASKS.AddCellAbove = new NotebookTask({
+  title: 'Add Cell Above',
+  keybindings: ['a'],
+  displayKeybinding: 'a',
+
+  callback() {
+    dispatcher.insertCell('javascript', 'above')
+    dispatcher.selectCell(getCellAboveSelectedId(), true)
+  },
+  keybindingCallback() {
+    if (isCommandMode()) {
+      dispatcher.insertCell('javascript', 'above')
+      dispatcher.selectCell(getCellAboveSelectedId(), true)
```

Comment:
Maybe replace with `this.callback()` to avoid code duplication here?  (I see above from some other tasks why both `callback`/`keybindingCallback` have to exist, unfortunately...)

Reply:
Yeah, I really want to figure out how to get rid of two callbacks. I HATE THAT WE HAVE TWO CALLBACKS and I won't sleep well until there is only one per `NotebookTask`.

Reply:
i actually think the two callbacks is a nice explicit pattern in cases where it's need -- i think mdboom just means that in this case it could be

```
keybindingCallback() {
    if (isCommandMode()) {
      this.callback()
    }
}
```

but i guess if we have a `NotebookCommandModeTask` wrapper class, it could by default just use `callback()` wrapped in `isCommandMode()` unless an expiicit `kebindingCallback` is provided. so maybe we can avoid two callbacks in many/most cases with wrapper classes. But i think having the flexibility and clarity of `keybingCallback` is great. I really like the approach you've come up with for this whole refactor :-)

Reply:
Yes, that's what I meant @bcolloran, but you said it better. :)

Reply:
Yeah, I suppose my issue with the two callbacks is almost always, with `myTask.keybindingCallback` what we really want is `callback` with an explicit `keybindingPrecondition`. The getter for `keybindingCallback` can remain the same API-wise, but what I want is just to say is

```javascript
const newTask = new NotebookTask({
   title: 'whatever',
   callback: ()=>{fireSomeAction()},
   keybinding:...,
   keybindingPrecondition: isCommandMode, // or some other function
})

```

And the API is such that `newTask.keybindingCallback` returns an anonymous function that looks like this:

```javascript
return function() {
   if (this.keybindingPrecondition(...arguments)) this.callback(...arguments);
}

```


These are small nits, and are not probably worth stopping the show for it, and at any rate your little code snippet actually is nice enough and reduces the complexity by ensuring things aren't out of sync. But this seems just about as explicit and feels a bit cleaner to me.

Reply:
@hamilton: I like what you propose as well.  I think it goes even further toward elegance.  But I'm happy to say "save for another PR" to avoid holding up this one.